omod/src/main/java/org/openmrs/web/controller/PortletController.java

```diff
@@ -195,43 +204,68 @@ public ModelAndView handleRequest(HttpServletRequest request, HttpServletRespons
+						
+						Person person = (Person) p;
+						List<Person> persons = Collections.singletonList(person);
+						
+						// Get most recent observations using limit parameter
+						List<Obs> paginatedObs = Context.getObsService().getObservations(persons, null, null, null, null,
+						    null, Collections.singletonList("obsDatetime desc"), limit, startIndex, null, null, false);
+						
+						model.put("limit", limit);
+						model.put("patientObs", paginatedObs);
+						
+						// Handle BMI calculation
 						Obs latestWeight = null;
 						Obs latestHeight = null;
 						String bmiAsString = "?";
+						
 						try {
 							String weightString = as.getGlobalProperty("concept.weight");
 							ConceptNumeric weightConcept = null;
 							if (StringUtils.hasLength(weightString)) {
 								weightConcept = cs.getConceptNumeric(GeneralUtils.getConcept(weightString).getConceptId());
 							}
+							
 							String heightString = as.getGlobalProperty("concept.height");
 							ConceptNumeric heightConcept = null;
 							if (StringUtils.hasLength(heightString)) {
 								heightConcept = cs.getConceptNumeric(GeneralUtils.getConcept(heightString).getConceptId());
 							}
-							for (Obs obs : patientObs) {
-								if (obs.getConcept().equals(weightConcept)) {
-									if (latestWeight == null
-									        || obs.getObsDatetime().compareTo(latestWeight.getObsDatetime()) > 0) {
-										latestWeight = obs;
-									}
-								} else if (obs.getConcept().equals(heightConcept)
-								        && (latestHeight == null || obs.getObsDatetime().compareTo(
-								            latestHeight.getObsDatetime()) > 0)) {
-									latestHeight = obs;
+							
+							if (weightConcept != null) {
+								List<Concept> weightQuestions = Collections.singletonList(weightConcept);
+								List<String> weightSort = Collections.singletonList("obsDatetime desc");
+								
+								List<Obs> weightObs = Context.getObsService().getObservations(persons, null,
+								    weightQuestions, null, null, null, weightSort, 1, null, null, null, false);
+								
+								if (!weightObs.isEmpty()) {
+									latestWeight = weightObs.get(0);
 								}
 							}
-							if (latestWeight != null) {
-								model.put("patientWeight", latestWeight);
-							}
-							if (latestHeight != null) {
-								model.put("patientHeight", latestHeight);
+							
+							if (heightConcept != null) {
+								List<Concept> heightQuestions = Collections.singletonList(heightConcept);
+								List<String> heightSort = Collections.singletonList("obsDatetime desc");
+								
+								List<Obs> heightObs = Context.getObsService().getObservations(persons, null,
+								    heightQuestions, null, null, null, heightSort, 1, null, null, null, false);
+								
+								if (!heightObs.isEmpty()) {
+									latestHeight = heightObs.get(0);
+								}
 							}
+							
 							if (latestWeight != null && latestHeight != null) {
+								model.put("patientWeight", latestWeight);
```

Comment:
What happens to the model when latestHeight is null but latestWeight is not null?

Reply:
No BMI in the Model: Since both latestWeight and latestHeight are required for BMI calculation, the model will not have patientBmi or patientBmiAsString if either is null. he model will only include the observation that is not null (patientWeight in this case). But since I will revert this, I think any changes like setting a default value or log warning wouldn't be necessary